google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,414 @@
+ *   }
+ *
+ *   for(CellChunk chunk : response.getChunksList()) {
+ *     boolean hasCompleteRow = stateMachine.handleChunk(chunk);
+ *     if (hasCompleteRow) {
+ *       MyRow row = stateMachine.consumeRow();
+ *
+ *       if (myRowAdapter.isScanMarkerRow(row) {
+ *         // do something with marker
+ *       } else {
+ *         // do something with row
+ *       }
+ *     }
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <RowT> The type of row the adapter will build
+ */
+final class StateMachine<RowT> {
+  private final RowBuilder<RowT> adapter;
+  private State currentState;
+  private ByteString lastCompleteRowKey;
+
+  // Track current cell attributes: protocol omits them when they are repeated
+  private ByteString rowKey;
+  private String familyName;
+  private ByteString qualifier;
+  private long timestamp;
+  private List<String> labels;
+  private long expectedCellSize;
+  private long remainingCellBytes;
+
+  private RowT completeRow;
+
+  /**
+   * Initialize a new state machine that's ready for a new row.
+   *
+   * @param adapter The adapter that will build the final row.
+   */
+  StateMachine(RowBuilder<RowT> adapter) {
+    this.adapter = adapter;
+    reset();
+  }
+
+  /**
+   * Feeds the last scanned serverside key into the state machine. The adapter will wrap this key in
+   * a special marker row that can be used downstream for efficient resume. It is the callers
+   * responsibility to eventually filter out this row.
+   *
+   * <dl>
+   *   <dt>Valid states:
+   *   <dd>{@link StateMachine#AWAITING_NEW_ROW}
+   *   <dt>Resulting states:
+   *   <dd>{@link StateMachine#AWAITING_ROW_CONSUME}
+   * </dl>
+   */
+  void handleLastScannedRow(ByteString key) {
+    try {
+      currentState = currentState.handleLastScannedRow(key);
+    } catch (RuntimeException e) {
+      currentState = null;
+      throw e;
+    }
+  }
+
+  /**
+   * Feeds a new chunk into the sate machine. If the chunk is invalid, the state machine will throw
+   * an exception and should be used for further input.
+   *
+   * <dl>
+   *   <dt>Valid states:
+   *   <dd>{@link StateMachine#AWAITING_NEW_ROW}
+   *   <dd>{@link StateMachine#AWAITING_NEW_CELL}
+   *   <dd>{@link StateMachine#AWAITING_CELL_VALUE}
+   *   <dt>Resulting states:
+   *   <dd>{@link StateMachine#AWAITING_NEW_CELL}
+   *   <dd>{@link StateMachine#AWAITING_CELL_VALUE}
+   *   <dd>{@link StateMachine#AWAITING_ROW_CONSUME}
+   * </dl>
+   *
+   * @param chunk The new chunk to process.
+   * @throws InvalidInputException When the chunk is not applicable to the current state.
+   * @throws IllegalStateException When the internal state is inconsistent
+   */
+  void handleChunk(CellChunk chunk) {
+    try {
+      currentState = currentState.handleChunk(chunk);
+    } catch (RuntimeException e) {
+      currentState = null;
+      throw e;
+    }
+  }
+
+  /**
+   * Returns the last completed row and transitions to awaiting a new row.
+   *
```

Comment:
Should every method have the valid/resulting states comment?

Reply:
Updated the docs